### PR TITLE
Add hook methods for tests and deployment in CompilerApi

### PR DESCRIPTION
**Pull request by cboulanger**
This lets you do 

```javascript
qx.Class.define("foo.CompilerApi", {
  extend: qx.tool.cli.api.CompilerApi,
  members: {
  
    beforeTests(command) {
      command.addTest(new qx.tool.cli.api.Test("fake test", async function() {
        console.log("# This is a fake test");
        this.setExitCode(0);
      })).setNeedsServer(false);
    },

    afterDeploy(data) {
      console.log("afterDeploy method called");
    },
```
without the event listener boilerplate.

### CODE REVIEW

File: source/class/qx/tool/cli/Cli.js [hunk 1]
```diff
@@ -170,7 +170,16 @@ qx.Class.define("qx.tool.cli.Cli", {
...
+    /**
+     * This is to notify the commands after loading the full args.
+     * The commands can overload special arg arguments here.
+     * e.g. Deploy will will overload the target.
+     */
+    __notifyCommand: function() {
+      this._compilerApi.getCommand().processArgs(this.argv);
+    },
+
     /**
```
Comment (cboulanger) [inline source/class/qx/tool/cli/Cli.js:183]: I don't know why this is in the PR, I didn't add it - it was in the source before!

File: source/class/qx/tool/cli/commands/Command.js [hunk 2]
```diff
@@ -60,14 +71,26 @@ qx.Class.define("qx.tool.cli.commands.Command", {
...
+     * e.g. Deploy will will overload the target.
+     *
+     * @param {*} argv : args to procvess
+     *
+     */
+    processArgs: function(argv) {
+       // Nothing
+    },
+
     /**
```
Comment (cboulanger) [inline source/class/qx/tool/cli/commands/Command.js:87]: This also isn't part of my commit - very strange!

File: source/class/qx/tool/utils/Utils.js [hunk 3]
```diff
@@ -231,6 +231,117 @@ qx.Class.define("qx.tool.utils.Utils", {
...
+        });
+      });
+    },
+
+    /**
+     * Node-8 compatible function to delete a directory recursively
+     * @param {String} dirname The name of the directory to delete
+     * @return {Promise}
+     */
+    async deleteRecursive(dirname) {
```
Comment (hkollmann) [inline source/class/qx/tool/utils/Utils.js:282]: Found in qx.tool.utils.files.Utils too. 
Is using rimraf there.

File: source/class/qx/tool/utils/Utils.js [hunk 4]
```diff
@@ -231,6 +231,117 @@ qx.Class.define("qx.tool.utils.Utils", {
...
+    },
+
+    /**
+     * Safely deletes a file asynchronously, without throwing
+     * if the file does not exist
+     * @param filename
+     * @return {Promise<void>}
+     */
+    async safeDelete(filename) {
+      try {
```
Comment (hkollmann) [inline source/class/qx/tool/utils/Utils.js:338]: Found in qx.tool.utils.files.Utils too,

**Comment by hkollmann** [comment]
To let the dog eat it's own food: Should'nt compile.js derived from compiler qx.compiler.CompilerApi? It's not an library.
And it would test your new implementation.

**Comment by cboulanger** [comment]
In the example, `foo.CompilerApi`extends `qx.tool.cli.api.CompilerApi` - but I think I don't fully understand what you mean - isn't compile.js just a node module?

**Comment by cboulanger** [comment]
Probably `qx.tool.utils.Utils.{mkpath, mkParentPath, makeParentDir, makeDirs}` should be moved to `qx.tool.utils.files.Utils`, then, to be consistent.

**Comment by hkollmann** [comment]
> Probably `qx.tool.utils.Utils.{mkpath, mkParentPath, makeParentDir, makeDirs}` should be moved to `qx.tool.utils.files.Utils`, then, to be consistent.

Good idea.

**Comment by hkollmann** [comment]
> In the example, `foo.CompilerApi`extends `qx.tool.cli.api.CompilerApi` - but I think I don't fully understand what you mean - isn't compile.js just a node module?

IMHO compile.js from compiler itselfs should be changed to use the CompilerApi

**Comment by cboulanger** [comment]
@hkollmann You are right - I noticed this too and wondered why it was `LibraryApi` when it really is a compiler thing. I can change it.

**Comment by hkollmann** [comment]
> @hkollmann You are right - I noticed this too and wondered why it was `LibraryApi` when it really is a compiler thing. I can change it.

Copy and paste without thinking about 😄

**Comment by cboulanger** [comment]
@hkollmann Now ok?

**Comment by cboulanger** [comment]
@johnspackman Can you have a look?

**Comment by cboulanger** [comment]
@johnspackman Can you review this, please?

**Comment by cboulanger** [comment]
Disabled travis for this repo, forcing merge.

**Pull request merged**